yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerProvisioningStrategy.java

```diff
@@ -58,7 +63,26 @@ public void setEnabled(boolean enabled) {
         final Label label = strategyState.getLabel();
         LoadStatisticsSnapshot snapshot = strategyState.getSnapshot();
 
-        for (DockerCloud dockerCloud : getDockerClouds()) {
+        //create a random list of docker clouds and prioritize idle clouds
+        List<DockerCloud> provisionClouds = null;
+        List<DockerCloud> availableClouds = getAvailableDockerClouds(label);
+        if (availableClouds.size() > 0) {
+            //select available clouds based on label which have potential capacity
+            LOG.debug("Picking from available clouds.");
+            provisionClouds = availableClouds;
+        } else {
+            //if there's no available clouds then fall back to original behavior
+            LOG.debug("Falling back to getting all clouds regardless of availability.");
+            provisionClouds = getDockerClouds();
+        }
+        //randomize the order of the DockerCloud list
+        Collections.shuffle(provisionClouds);
+        //sort by least loaded DockerCloud (i.e. fewest provisioned slaves)
+        Collections.sort(provisionClouds, new DockerCloudLoadComparator());
```

Comment:
Why shuffle is needed if you are sorting?

Reply:
The shuffle happens to randomize the order.  The sorting is sorted based on least load.  So clouds with the same load will be together but randomly ordered.